nexus/reconfigurator/rendezvous/src/debug_dataset.rs

```diff
@@ -0,0 +1,297 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+//! Recording debug datasets in their rendezvous table
+
+use anyhow::Context;
+use nexus_db_queries::context::OpContext;
+use nexus_db_queries::db::model::RendezvousDebugDataset;
+use nexus_db_queries::db::DataStore;
+use nexus_types::deployment::BlueprintDatasetConfig;
+use nexus_types::deployment::BlueprintDatasetDisposition;
+use omicron_common::api::internal::shared::DatasetKind;
+use omicron_uuid_kinds::BlueprintUuid;
+use omicron_uuid_kinds::DatasetUuid;
+use slog::info;
+use std::collections::BTreeSet;
+
+pub(crate) async fn reconcile_debug_datasets(
+    opctx: &OpContext,
+    datastore: &DataStore,
+    blueprint_id: BlueprintUuid,
+    blueprint_datasets: impl Iterator<Item = &BlueprintDatasetConfig>,
+    inventory_datasets: &BTreeSet<DatasetUuid>,
+) -> anyhow::Result<()> {
+    // We expect basically all executions of this task to do nothing: we're
+    // activated periodically, and only do work when a dataset has been
+    // newly-added or newly-expunged. For the newly-added case, we can first
+    // fetch all the existing datasets and then avoid trying to spuriously
+    // insert them again.
+    //
+    // This is a minor performance optimization. If we removed this fetch, the
+    // code below would still be correct, but it would issue a bunch of
+    // do-nothing inserts for already-existing datasets.
+    let existing_datasets = datastore
+        .debug_dataset_list_all_batched(opctx)
+        .await
+        .context("failed to list all debug datasets")?
+        .into_iter()
+        .map(|d| d.id())
+        .collect::<BTreeSet<_>>();
+
+    // We want to insert any in-service datasets (according to the blueprint)
+    // that are also present in `inventory_datasets` but that are not already
+    // present in the database (described by `existing_datasets`).
+    let datasets_to_insert = inventory_datasets
```

Comment:
```suggestion
    let new_inventory_datasets = inventory_datasets
```

This is a nitpick, but I think it's worthwhile. We don't necessarily want to insert these debug datasets unless they're also in the blueprint as in-service.

Reply:
I reworked this in c9c846379 to not even construct this intermediate set.